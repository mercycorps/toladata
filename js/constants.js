function getGlobal(key) {
    if ((typeof JS_GLOBALS !== 'undefined')  && JS_GLOBALS.hasOwnProperty(key)) {
        return JS_GLOBALS[key];
    }
    return null;
}


/**
 * IPTT Constants:
 */
const BLANK_LABEL = '---------';
const BLANK_OPTION = {
    value: null,
    label: BLANK_LABEL
};
const BLANK_TABLE_CELL = 'â€”';


const TVA = 1;
const TIMEPERIODS = 2;

const TIME_AWARE_FREQUENCIES = [3, 4, 5, 6, 7];
const IRREGULAR_FREQUENCIES = [1, 2];

const TVA_FREQUENCY_LABELS = Object.freeze(
    {
        1: gettext("Life of Program (LoP) only"),
        2: gettext("Midline and endline"),
        3: gettext("Annual"),
        4: gettext("Semi-annual"),
        5: gettext("Tri-annual"),
        6: gettext("Quarterly"),
        7: gettext("Monthly")
    }
);

const TIMEPERIODS_FREQUENCY_LABELS = Object.freeze(
    {
        3: gettext("Years"),
        4: gettext("Semi-annual periods"),
        5: gettext("Tri-annual periods"),
        6: gettext("Quarters"),
        7: gettext("Months")
    }
);

export { BLANK_OPTION, BLANK_LABEL, BLANK_TABLE_CELL, TVA, TIMEPERIODS, TIME_AWARE_FREQUENCIES,
         IRREGULAR_FREQUENCIES, TVA_FREQUENCY_LABELS, TIMEPERIODS_FREQUENCY_LABELS };

const GROUP_BY_CHAIN = 1;
const GROUP_BY_LEVEL = 2;

export { GROUP_BY_CHAIN, GROUP_BY_LEVEL };

const _gettext = (typeof gettext !== 'undefined') ?  gettext : (s) => s;

function getPeriodLabels() {
    return {
        targetperiodLabels: {
            1: _gettext("Life of Program (LoP) only"),
            3: _gettext("Annual"),
            2: _gettext("Midline and endline"),
            5: _gettext("Tri-annual"),
            4: _gettext("Semi-annual"),
            7: _gettext("Monthly"),
            6: _gettext("Quarterly")
        },
        timeperiodLabels: {
            3: _gettext("Years"),
            5: _gettext("Tri-annual periods"),
            4: _gettext("Semi-annual periods"),
            7: _gettext("Months"),
            6: _gettext("Quarters")
        }
    };
}

export {getPeriodLabels};

export const STATUS_CODES = {
    NO_INDICATOR_IN_UPDATE: 1
};

export const IndicatorFilterType = Object.freeze({
    noFilter: 0,
    missingTarget: 1,
    missingResults: 2,
    missingEvidence: 3,

    aboveTarget: 5,
    belowTarget: 6,
    onTarget: 7,
    nonReporting: 8
});

export const RFC_OPTIONS = getGlobal('reason_for_change_options') ?? [];