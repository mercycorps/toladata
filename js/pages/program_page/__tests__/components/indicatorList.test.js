import '../../../../test_helpers/django_i18n_stubs';
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, render } from 'enzyme';
import IndicatorList, { StatusHeader, IndicatorFilter, IndicatorListTable } from '../../components/indicator_list';
import { IndicatorFilterType } from '../../../../constants';
import indicatorResultData from './fixtures/resultsTableIndicators.json';


import eventBus from '../../../../eventbus';

jest.mock('../../../../eventbus');

describe("Indicator List elements", () => {
    describe("Status Header component", () => {
        let props = {
            indicatorCount: 10,
            programId: 144,
            currentIndicatorFilter: null,
            filterApplied: false,
            readonly: false
        };
        it("renders without crashing", () => {
            const component = renderer.create(
                <StatusHeader {...props} />
            );
            expect(component.toJSON()).toMatchSnapshot();
            const wrapper = shallow(
                <StatusHeader {...props} />
            );
            expect(wrapper.find('#show-all-indicators').exists()).toBeFalsy();
        });
        it("renders read only", () => {
            const component = renderer.create(
                <StatusHeader {...props} readonly={true} />
            );
            expect(component.toJSON()).toMatchSnapshot();
            const wrapper = shallow(
                <StatusHeader {...props} readonly={true} />
            );
            expect(wrapper.find('div > div').children().length).toBe(0);
            expect(wrapper.find('#show-all-indicators').exists()).toBeFalsy();
        });
        it("renders filter applied", () => {
            const component = renderer.create(
                <StatusHeader {...props} filterApplied={true} />
            );
            expect(component.toJSON()).toMatchSnapshot();
            const wrapper = shallow(
                <StatusHeader {...props} filterApplied={true} />
            );
            expect(wrapper.find('#show-all-indicators').exists()).toBeTruthy();
        });
        it("calls for filter title", () => {
            const component = renderer.create(
                <StatusHeader {...props} filterApplied={true}
                indicatorCount={15}
                currentIndicatorFilter={IndicatorFilterType.aboveTarget} />
            );
            expect(component.toJSON()).toMatchSnapshot();
            const wrapper = shallow(
                <StatusHeader {...props} filterApplied={true}
                indicatorCount={15}
                currentIndicatorFilter={IndicatorFilterType.aboveTarget} />
            );
            //correct title text:
            expect(wrapper.find('div > h3 > span').text()).toEqual("%s indicator is >15% above target");
            // show all link is displayed:
            expect(wrapper.find('#show-all-indicators').exists()).toBeTruthy();
            expect(wrapper.find('#show-all-indicators').shallow().text()).toBe("Show all indicators");
            //text was translated:
            expect(window.gettext).toHaveBeenCalledWith('Show all indicators');
        });
    });
    describe("Indicator Filter component", () => {
        var props;
        beforeEach(() => {
            props = {
                rootStore: {
                    allIndicators: []
                },
                uiStore: {
                    resultChainFilterLabel: 'label',
                    selectedGroupByOption: null,
                    selectedIndicatorId: null,
                    groupByOptions: []
                }
            };
        });
        it("renders without crashing", () => {
            const component = renderer.create(
                <IndicatorFilter {...props} />
            );
            expect(component.toJSON()).not.toBeUndefined();
        });
        it("renders indicator options", () => {
            const indicatorOptions = [
                {pk: 1, name: 'one'},
                {pk: 2, name: 'two'}
            ];
            const expectedOptions = [
                {value: 1, label: 'one'},
                {value: 2, label: 'two'}
            ];
            props.rootStore.allIndicators = indicatorOptions;
            props.uiStore.selectedIndicatorId = 1;
            props.uiStore.setSelectedIndicatorId = jest.fn();
            const wrapper = shallow(
                <IndicatorFilter {...props} />
            );
            const indicatorSelectDiv = wrapper.find('.form-group').at(0).shallow();
            expect(indicatorSelectDiv.find('label').first().text()).toEqual("Find an indicator:");
            expect(window.gettext).toHaveBeenCalledWith("Find an indicator:");
            const indicatorSelect =  indicatorSelectDiv.find('div > div').childAt(0).shallow();
            expect(indicatorSelect.props().options).toEqual(expectedOptions);
            expect(indicatorSelect.props().value).toEqual(expectedOptions[0]);
            indicatorSelect.simulate('change', expectedOptions[1]);
            expect(eventBus.emit).toHaveBeenCalledWith('nav-select-indicator-to-filter', 2);
            //expect(props.uiStore.setSelectedIndicatorId).toHaveBeenCalledWith(2);
        });

        it("renders group by options", () => {
            const groupByOptions = [
                {value: 1, label: 'chain'},
                {value: 2, label: 'level'}
            ];
            props.uiStore.groupByOptions = groupByOptions;
            props.uiStore.selectedGroupByOption = 1;
            props.uiStore.setGroupBy = jest.fn();
            const wrapper = shallow(
                <IndicatorFilter {...props} />
            );
            expect(wrapper.find('.form-group').length).toBe(2);
            const groupBySelectDiv = wrapper.find('.form-group').at(1).shallow();
            expect(groupBySelectDiv.find('label').first().text()).toEqual("Group indicators:");
            expect(window.gettext).toHaveBeenCalledWith("Group indicators:");
            const groupBySelect =  groupBySelectDiv.find('div > div').childAt(0).shallow();
            expect(groupBySelect.props().options).toEqual(groupByOptions);
            expect(groupBySelect.props().value).toEqual(1);
            groupBySelect.simulate('change', groupByOptions[1]);
            expect(props.uiStore.setGroupBy).toHaveBeenCalledWith(2);
        });
    });
    describe("Indicator List Table component", () => {
        var props;
        beforeEach(() => {
            props = {
                indicators: [],
                program: {
                    resultsMap: new Map()
                }
            };
        });
        it("renders without crashing", () => {
            const component = renderer.create(
                <IndicatorListTable {...props} />
            );
            expect(component.toJSON()).toMatchSnapshot();
        });
        it("appropriately displays a disassociated result warning", () => {
            props.indicators.push(indicatorResultData.unassignedResultsWithTargets);
            props.program.isExpanded = x => false;
            let component = render(<IndicatorListTable {...props} />);
            expect(component.html()).toContain("Results unassigned to targets");
            props.program.isExpanded = x => true;
            component = render(<IndicatorListTable {...props} />);
            expect(component.html()).toContain("Results unassigned to targets");
            props.indicators[0].periodicTargets = []
            component = render(<IndicatorListTable {...props} />);
            expect(component.html()).toEqual(expect.not.stringContaining("Results unassigned to targets"));
        });
    });
});

describe("Indicator List component", () => {
    it.skip("nothing yet", () => {});
});
