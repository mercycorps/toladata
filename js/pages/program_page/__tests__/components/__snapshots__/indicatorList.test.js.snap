// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indicator List elements Indicator List Table component renders without crashing 1`] = `
<table
  className="table indicators-list"
>
  <thead>
    <tr
      className="table-header"
    >
      <th
        className=""
        id="id_indicator_name_col_header"
      >
        Indicator
      </th>
      <th
        className=""
        id="id_indicator_buttons_col_header"
      >
        Â 
      </th>
      <th
        className=""
        id="id_indicator_unit_col_header"
      >
        Unit of measure
      </th>
      <th
        className="text-right"
        id="id_indicator_baseline_col_header"
      >
        Baseline
      </th>
      <th
        className="text-right"
        id="id_indicator_target_col_header"
      >
        Target
      </th>
    </tr>
  </thead>
  <tbody />
</table>
`;

exports[`Indicator List elements Status Header component calls for filter title 1`] = `
<div
  className="indicators-list__header"
>
  <h3
    className="no-bold"
  >
    <span
      id="indicators-list-title"
    >
      %s indicator is &gt;15% above target
    </span>
    <small
      className="ml-2 text-medium-dark text-nowrap"
    >
      |
      <a
        className="btn btn-sm btn-link btn-inline ml-2"
        href="#"
        id="show-all-indicators"
        onClick={[Function]}
      >
        Show all indicators
      </a>
    </small>
  </h3>
  <div>
    <button
      className="btn btn-sm btn-link px-0 btn-add text-nowrap"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-plus-circle"
      />
       
      Add indicator
    </button>
  </div>
</div>
`;

exports[`Indicator List elements Status Header component renders filter applied 1`] = `
<div
  className="indicators-list__header"
>
  <h3
    className="no-bold"
  >
    <span
      id="indicators-list-title"
    >
      %s indicator
    </span>
    <small
      className="ml-2 text-medium-dark text-nowrap"
    >
      |
      <a
        className="btn btn-sm btn-link btn-inline ml-2"
        href="#"
        id="show-all-indicators"
        onClick={[Function]}
      >
        Show all indicators
      </a>
    </small>
  </h3>
  <div>
    <button
      className="btn btn-sm btn-link px-0 btn-add text-nowrap"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-plus-circle"
      />
       
      Add indicator
    </button>
  </div>
</div>
`;

exports[`Indicator List elements Status Header component renders read only 1`] = `
<div
  className="indicators-list__header"
>
  <h3
    className="no-bold"
  >
    <span
      id="indicators-list-title"
    >
      %s indicator
    </span>
  </h3>
  <div />
</div>
`;

exports[`Indicator List elements Status Header component renders without crashing 1`] = `
<div
  className="indicators-list__header"
>
  <h3
    className="no-bold"
  >
    <span
      id="indicators-list-title"
    >
      %s indicator
    </span>
  </h3>
  <div>
    <button
      className="btn btn-sm btn-link px-0 btn-add text-nowrap"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-plus-circle"
      />
       
      Add indicator
    </button>
  </div>
</div>
`;
